core/pkg/store/store.go

```diff
@@ -131,146 +213,176 @@ func (f *Store) String() (string, error) {
+		newFlag.Source = source
+		newFlag.Priority = priority
+		newFlag.Metadata = patchMetadata(metadata, newFlag.Metadata)
+
+		// flagSetId defaults to a UUID generated at startup to make our quires isomorphic
+		flagSetId := nilFlagSetId
+		// flagSetId is inherited from the set, but can be overridden by the flag
+		setFlagSetId, ok := newFlag.Metadata["flagSetId"].(string)
 		if ok {
-			if !f.hasPriority(storedFlag.Source, source) {
-				f.logger.Debug(
-					fmt.Sprintf(
-						"not merging: flag %s from source %s does not have priority over %s",
-						key, source, storedFlag.Source,
-					),
-				)
-				continue
-			}
-			if reflect.DeepEqual(storedFlag, newFlag) {
-				continue
-			}
+			flagSetId = setFlagSetId
 		}
-		if !ok {
-			notifications[key] = map[string]interface{}{
-				"type":   string(model.NotificationCreate),
-				"source": source,
-			}
-		} else {
-			notifications[key] = map[string]interface{}{
-				"type":   string(model.NotificationUpdate),
-				"source": source,
+		newFlag.FlagSetId = flagSetId
+
+		raw, err := txn.First(flagsTable, keySourceCompoundIndex, key, source)
+		if err != nil {
+			s.logger.Error(fmt.Sprintf("unable to get flag %s from source %s: %v", key, source, err))
+			continue
+		}
+		oldFlag, ok := raw.(model.Flag)
+		// If we already have a flag with the same key and source, we need to check if it has the same flagSetId
+		if ok {
+			if oldFlag.FlagSetId != newFlag.FlagSetId {
+				// If the flagSetId is different, we need to delete the, since flagSetId+key represents the primary index, and it's now been changed.
+				// Yhis is important especially for clients listening to flagSetId changes, as they expect the flag to be removed from the set in this case.
+				_, err = txn.DeleteAll(flagsTable, idIndex, oldFlag.FlagSetId, key)
+				if err != nil {
+					s.logger.Error(fmt.Sprintf("unable to delete flags with key %s and flagSetId %s: %v", key, oldFlag.FlagSetId, err))
+					continue
+				}
 			}
 		}
 		// Store the new version of the flag
-		err := txn.Insert("flags", newFlag)
+		s.logger.Debug(fmt.Sprintf("storing flag: %v", newFlag))
+		err = txn.Insert(flagsTable, newFlag)
 		if err != nil {
-			f.logger.Error(fmt.Sprintf("unable to insert flag %s: %v", key, err))
+			s.logger.Error(fmt.Sprintf("unable to insert flag %s: %v", key, err))
 			continue
 		}
 	}
 
 	txn.Commit()
-	return notifications, resyncRequired
+	return notifications.NewFromFlags(oldFlags, flags), resyncRequired
 }
 
-func (f *Store) GetMetadataForSource(source string) model.Metadata {
-	f.mx.RLock()
-	defer f.mx.RUnlock()
-	perSource, ok := f.MetadataPerSource[source]
-	if ok && perSource != nil {
-		return maps.Clone(perSource)
+// Watch the result-set of a selector for changes, sending updates to the watcher channel.
+func (s *Store) Watch(ctx context.Context, selector Selector, watcher chan FlagQueryResult) {
+	go func() {
+		for {
+			ws := memdb.NewWatchSet()
+			it, err := s.selectOrAll(selector)
```

Comment:
I would log at ERROR level and return if `err != nil`

Reply:
Good catch, added a return and close of the `watcher` chan.